### PR TITLE
Allow only 10% of connections from non-bloxroute gateway peers

### PR DESCRIPTION
**Pull request by sergeyi23**

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -1562,6 +1562,28 @@ void CConnman::WakeMessageHandler() {
     condMsgProc.notify_one();
 }
 
+bool CConnman::BxHasSlotForPeer(std::string strSubVer) {
+    {
+        LOCK(cs_vNodes);
+
+        int bxConnectionsCount = 0;
+        for (const CNode *pnode : vNodes) {
+            if (pnode->strSubVer.find(BLOXROUTE_GATEWAY_AGENT_NAME) == 0) {
```
Comment (avrahamblx) [inline src/net.cpp]: is `nullptr` check make sense here?

File: src/net.cpp [hunk 2]
```diff
@@ -1562,6 +1562,28 @@ void CConnman::WakeMessageHandler() {
...
+                bxConnectionsCount++;
+            }
+        }
+
+        int availableConnections;
+        if (strSubVer.find(BLOXROUTE_GATEWAY_AGENT_NAME) == 0) {
+            availableConnections = round((double)nMaxConnections * 0.9) - bxConnectionsCount;
```
Comment (avrahamblx) [inline src/net.cpp]: does checking for overflows make sense here?

File: src/net.cpp [hunk 3]
```diff
@@ -1562,6 +1562,28 @@ void CConnman::WakeMessageHandler() {
...
+        } else {
+            availableConnections = round((double)nMaxConnections * 0.1) - (vNodes.size() - bxConnectionsCount);
```
Comment (avrahamblx) [inline src/net.cpp]: same question here?

File: src/net.h [hunk 4]
```diff
@@ -94,6 +94,8 @@ static const size_t DEFAULT_MAXSENDBUFFER = 1 * 1000;
 // NOTE: When adjusting this, update rpcnet:setban's help ("24h")
 static const unsigned int DEFAULT_MISBEHAVING_BANTIME = 60 * 60 * 24;
 
+static const std::string BLOXROUTE_GATEWAY_AGENT_NAME = "bloXroute Gateway";
```
Comment (avrahamblx) [inline src/net.h]: depends on what C++ version you using `constexpr` may give better performance.

File: src/net.cpp [hunk 5]
```diff
@@ -1562,6 +1562,28 @@ void CConnman::WakeMessageHandler() {
     condMsgProc.notify_one();
 }
 
+bool CConnman::BxHasSlotForPeer(std::string strSubVer) {
```
Comment (avrahamblx) [inline src/net.cpp]: any reason of passing `strSubVer` by value?

File: src/net.h [hunk 6]
```diff
@@ -321,6 +323,8 @@ class CConnman {
 
     void WakeMessageHandler();
 
+    bool BxHasSlotForPeer(std::string serviceFlags);
```
Comment (avrahamblx) [inline src/net.h]: argument name doesn't match the implementation one!

File: src/net.cpp (hunk 3 above)
Comment (Uri-Margalit-bloXroute) [reply src/net.cpp]: doesn't `vNodes.size()` include `bxConnectionsCount`?

File: src/net.cpp (hunk 5 above)
Comment (sergeyi23) [reply src/net.cpp]: Fixed

File: src/net.cpp [hunk 7]
```diff
@@ -1562,6 +1562,28 @@ void CConnman::WakeMessageHandler() {
...
+    {
+        LOCK(cs_vNodes);
+
+        int bxConnectionsCount = 0;
+        for (const CNode *pnode : vNodes) {
+            if (pnode->strSubVer.find(BLOXROUTE_GATEWAY_AGENT_NAME) == 0) {
+                bxConnectionsCount++;
```
[the remark this answers is not in the document]
Comment (sergeyi23) [reply src/net.cpp]: Fixed

File: src/net.h (hunk 4 above)
Comment (sergeyi23) [reply src/net.h]: Fixed

File: src/net.h (hunk 6 above)
Comment (sergeyi23) [reply src/net.h]: Fixed

**Pull request merged**